### ISSUE TITLE
cloudformation deploy exits without useful message

### ISSUE THREAD

**Issue by byumark** [awslabs/serverless-application-model#58]
After running the `aws cloudformation package` command, I then run deploy and get this:

```
aws cloudformation deploy --template-file new_template.yml --stack-name stack_name --capabilities CAPABILITY_IAM --profile profile_name
Waiting for changeset to be created..

'Status'
``` 

It exits at that point. It gets stuck on the first event:
"ResourceStatus": "REVIEW_IN_PROGRESS", 

I am using the code from this example:
https://github.com/awslabs/serverless-application-model/tree/master/examples/2016-10-31/hello_world

aws-cli/1.11.36 Python/2.7.10 Darwin/16.3.0 botocore/1.4.93

**Comment by byumark** [comment]
Just an update:
Hhaving only this cloudformation action allowed `cloudformation:DescribeStacks` will give this error:

`An error occurred (AccessDenied) when calling the CreateChangeSet operation: User: arn:aws:iam::account_id:user/user_name is not authorized to perform: cloudformation:CreateChangeSet`

So when I allow `cloudformation:CreateChangeSet` by adding it to the policy, then I get the exit with this message:

`Waiting for changeset to be created..

'Status'`

I'm guessing I need to allow more actions but there is no error message telling me what I'm missing.

**Comment by vikrambhatt** [comment]
Does it complete if you call deploy with '--no-execute-changeset' option?
If yes, then you might be missing `cloudformation:ExecuteChangeSet` permission.

**Comment by byumark** [comment]
aws cloudformation deploy --template-file new_template.yml --stack-name stack_name --capabilities CAPABILITY_IAM --profile profile_name --no-execute-changeset

This produces the same result.

Allowing cloudformation:* in the policy causes it to work.

**Comment by sanathkr** [comment]
`deploy` command calls CreateChangeSet API followed by ExecuteChangeSet API. You should have both permissions in your policy for this to work. Of course, you should have permissions to create/manage the resources specified in your template.

My guess is that the CLI does not handle exceptions when calling ExecuteChangeSet. Needs more investigation

**Comment by syang** [comment]
@sanathkr I met a problem when experimenting your code and here is some info (looks like the lambda function creation failed)


$ aws cloudformation describe-stack-events --stack-name api-backend-serverless-2 
{
    "StackEvents": [
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "6ee2d650-30f5-11e7-a17e-5081e7e8c4ad", 
            "ResourceStatus": "ROLLBACK_COMPLETE", 
            "ResourceType": "AWS::CloudFormation::Stack", 
            "Timestamp": "2017-05-04T18:13:54.309Z", 
            "StackName": "api-backend-serverless-2", 
            "PhysicalResourceId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "LogicalResourceId": "api-backend-serverless-2"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "Table-DELETE_COMPLETE-2017-05-04T18:13:53.299Z", 
            "ResourceStatus": "DELETE_COMPLETE", 
            "ResourceType": "AWS::DynamoDB::Table", 
            "Timestamp": "2017-05-04T18:13:53.299Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"AttributeDefinitions\":[{\"AttributeType\":\"S\",\"AttributeName\":\"id\"}],\"ProvisionedThroughput\":{\"WriteCapacityUnits\":\"5\",\"ReadCapacityUnits\":\"5\"},\"KeySchema\":[{\"KeyType\":\"HASH\",\"AttributeName\":\"id\"}]}\n", 
            "PhysicalResourceId": "api-backend-serverless-2-Table-1DYSXTD86QNF9", 
            "LogicalResourceId": "Table"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "Table-DELETE_IN_PROGRESS-2017-05-04T18:13:22.158Z", 
            "ResourceStatus": "DELETE_IN_PROGRESS", 
            "ResourceType": "AWS::DynamoDB::Table", 
            "Timestamp": "2017-05-04T18:13:22.158Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"AttributeDefinitions\":[{\"AttributeType\":\"S\",\"AttributeName\":\"id\"}],\"ProvisionedThroughput\":{\"WriteCapacityUnits\":\"5\",\"ReadCapacityUnits\":\"5\"},\"KeySchema\":[{\"KeyType\":\"HASH\",\"AttributeName\":\"id\"}]}\n", 
            "PhysicalResourceId": "api-backend-serverless-2-Table-1DYSXTD86QNF9", 
            "LogicalResourceId": "Table"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "GetFunctionRole-DELETE_COMPLETE-2017-05-04T18:13:22.000Z", 
            "ResourceStatus": "DELETE_COMPLETE", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:22Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBReadOnlyAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "GetFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "DeleteFunctionRole-DELETE_COMPLETE-2017-05-04T18:13:21.904Z", 
            "ResourceStatus": "DELETE_COMPLETE", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:21.904Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "DeleteFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "PutFunctionRole-DELETE_COMPLETE-2017-05-04T18:13:21.878Z", 
            "ResourceStatus": "DELETE_COMPLETE", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:21.878Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "PutFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "5833ccc0-30f5-11e7-8617-500c2171ae62", 
            "ResourceStatus": "ROLLBACK_IN_PROGRESS", 
            "ResourceType": "AWS::CloudFormation::Stack", 
            "Timestamp": "2017-05-04T18:13:16.258Z", 
            "ResourceStatusReason": "The following resource(s) failed to create: [GetFunctionRole, Table, PutFunctionRole, DeleteFunctionRole]. . Rollback requested by user.", 
            "StackName": "api-backend-serverless-2", 
            "PhysicalResourceId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "LogicalResourceId": "api-backend-serverless-2"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "Table-CREATE_FAILED-2017-05-04T18:13:15.229Z", 
            "ResourceStatus": "CREATE_FAILED", 
            "ResourceType": "AWS::DynamoDB::Table", 
            "Timestamp": "2017-05-04T18:13:15.229Z", 
            "ResourceStatusReason": "Resource creation cancelled", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"AttributeDefinitions\":[{\"AttributeType\":\"S\",\"AttributeName\":\"id\"}],\"ProvisionedThroughput\":{\"WriteCapacityUnits\":\"5\",\"ReadCapacityUnits\":\"5\"},\"KeySchema\":[{\"KeyType\":\"HASH\",\"AttributeName\":\"id\"}]}\n", 
            "PhysicalResourceId": "api-backend-serverless-2-Table-1DYSXTD86QNF9", 
            "LogicalResourceId": "Table"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "GetFunctionRole-CREATE_FAILED-2017-05-04T18:13:14.890Z", 
            "ResourceStatus": "CREATE_FAILED", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:14.890Z", 
            "ResourceStatusReason": "API: iam:CreateRole User: arn:aws:iam::159377261049:user/serverlessAppAdmin is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::159377261049:role/api-backend-serverless-2-GetFunctionRole-OMYULN7H4E5F", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBReadOnlyAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "GetFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "DeleteFunctionRole-CREATE_FAILED-2017-05-04T18:13:14.448Z", 
            "ResourceStatus": "CREATE_FAILED", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:14.448Z", 
            "ResourceStatusReason": "API: iam:CreateRole User: arn:aws:iam::159377261049:user/serverlessAppAdmin is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::159377261049:role/api-backend-serverless-2-DeleteFunctionRole-FMTRWG9V61R6", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "DeleteFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "PutFunctionRole-CREATE_FAILED-2017-05-04T18:13:14.367Z", 
            "ResourceStatus": "CREATE_FAILED", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:14.367Z", 
            "ResourceStatusReason": "API: iam:CreateRole User: arn:aws:iam::159377261049:user/serverlessAppAdmin is not authorized to perform: iam:CreateRole on resource: arn:aws:iam::159377261049:role/api-backend-serverless-2-PutFunctionRole-1L95FWNFPVAK3", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "PutFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "GetFunctionRole-CREATE_IN_PROGRESS-2017-05-04T18:13:13.992Z", 
            "ResourceStatus": "CREATE_IN_PROGRESS", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:13.992Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBReadOnlyAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "GetFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "Table-CREATE_IN_PROGRESS-2017-05-04T18:13:13.880Z", 
            "ResourceStatus": "CREATE_IN_PROGRESS", 
            "ResourceType": "AWS::DynamoDB::Table", 
            "Timestamp": "2017-05-04T18:13:13.880Z", 
            "ResourceStatusReason": "Resource creation Initiated", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"AttributeDefinitions\":[{\"AttributeType\":\"S\",\"AttributeName\":\"id\"}],\"ProvisionedThroughput\":{\"WriteCapacityUnits\":\"5\",\"ReadCapacityUnits\":\"5\"},\"KeySchema\":[{\"KeyType\":\"HASH\",\"AttributeName\":\"id\"}]}\n", 
            "PhysicalResourceId": "api-backend-serverless-2-Table-1DYSXTD86QNF9", 
            "LogicalResourceId": "Table"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "PutFunctionRole-CREATE_IN_PROGRESS-2017-05-04T18:13:13.662Z", 
            "ResourceStatus": "CREATE_IN_PROGRESS", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:13.662Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "PutFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "DeleteFunctionRole-CREATE_IN_PROGRESS-2017-05-04T18:13:13.647Z", 
            "ResourceStatus": "CREATE_IN_PROGRESS", 
            "ResourceType": "AWS::IAM::Role", 
            "Timestamp": "2017-05-04T18:13:13.647Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"ManagedPolicyArns\":[\"arn:aws:iam::aws:policy/AmazonDynamoDBFullAccess\",\"arn:aws:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole\"],\"AssumeRolePolicyDocument\":{\"Version\":\"2012-10-17\",\"Statement\":[{\"Action\":[\"sts:AssumeRole\"],\"Effect\":\"Allow\",\"Principal\":{\"Service\":[\"lambda.amazonaws.com\"]}}]}}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "DeleteFunctionRole"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "Table-CREATE_IN_PROGRESS-2017-05-04T18:13:13.468Z", 
            "ResourceStatus": "CREATE_IN_PROGRESS", 
            "ResourceType": "AWS::DynamoDB::Table", 
            "Timestamp": "2017-05-04T18:13:13.468Z", 
            "StackName": "api-backend-serverless-2", 
            "ResourceProperties": "{\"AttributeDefinitions\":[{\"AttributeType\":\"S\",\"AttributeName\":\"id\"}],\"ProvisionedThroughput\":{\"WriteCapacityUnits\":\"5\",\"ReadCapacityUnits\":\"5\"},\"KeySchema\":[{\"KeyType\":\"HASH\",\"AttributeName\":\"id\"}]}\n", 
            "PhysicalResourceId": "", 
            "LogicalResourceId": "Table"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "53c85c50-30f5-11e7-b996-504dcd69ef99", 
            "ResourceStatus": "CREATE_IN_PROGRESS", 
            "ResourceType": "AWS::CloudFormation::Stack", 
            "Timestamp": "2017-05-04T18:13:08.846Z", 
            "ResourceStatusReason": "User Initiated", 
            "StackName": "api-backend-serverless-2", 
            "PhysicalResourceId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "LogicalResourceId": "api-backend-serverless-2"
        }, 
        {
            "StackId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "EventId": "4d5b9da0-30f5-11e7-9a15-500c21fb2c29", 
            "ResourceStatus": "REVIEW_IN_PROGRESS", 
            "ResourceType": "AWS::CloudFormation::Stack", 
            "Timestamp": "2017-05-04T18:12:58.238Z", 
            "ResourceStatusReason": "User Initiated", 
            "StackName": "api-backend-serverless-2", 
            "PhysicalResourceId": "arn:aws:cloudformation:us-west-1:159377261049:stack/api-backend-serverless-2/4d5c60f0-30f5-11e7-9a15-500c21fb2c29", 
            "LogicalResourceId": "api-backend-serverless-2"
        }
    ]
}

**Comment by syang** [comment]
@sanathkr I copied and pasted your api_gateway code, and the deployed failed. Here is the detailed info, and hopefully this gives you some info to help out.

1) First, I **replaced my specific s3 bucket name to <s3 bucket>** in the template.yaml
`AWSTemplateFormatVersion: '2010-09-09'
Transform: AWS::Serverless-2016-10-31
Description: Simple CRUD webservice. State is stored in a SimpleTable (DynamoDB) resource.
Resources:
  GetFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.get
      Runtime: nodejs4.3
      CodeUri: s3://serverless-asset/api_backend.zip
      Policies: AmazonDynamoDBReadOnlyAccess
      Environment:
        Variables:
          TABLE_NAME: !Ref Table
      Events:
        GetResource:
          Type: Api
          Properties:
            Path: /resource/{resourceId}
            Method: get

  PutFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.put
      Runtime: nodejs4.3
      CodeUri: s3://serverless-asset/api_backend.zip
      Policies: AmazonDynamoDBFullAccess
      Environment:
        Variables:
          TABLE_NAME: !Ref Table
      Events:
        PutResource:
          Type: Api
          Properties:
            Path: /resource/{resourceId}
            Method: put

  DeleteFunction:
    Type: AWS::Serverless::Function
    Properties:
      Handler: index.delete
      Runtime: nodejs4.3
      CodeUri: s3://serverless-asset/api_backend.zip
      Policies: AmazonDynamoDBFullAccess
      Environment:
        Variables:
          TABLE_NAME: !Ref Table
      Events:
        DeleteResource:
          Type: Api
          Properties:
            Path: /resource/{resourceId}
            Method: delete

  Table:
    Type: AWS::Serverless::SimpleTable
`
2) then I use the following command to package and deploy the stack
$aws cloudformation package --template-file template.yaml --output-template-file new_template.yaml --s3-bucket s3://serverless-asset

$aws cloudformation deploy --template-file ./serverless/example_api_backend/new_template.yaml --stack-name serverless-api-backend-test-4 --capabilities CAPABILITY_IAM

3) Then awscli told me it failed, and I searched what action/step failed

 $ aws cloudformation describe-stack-events --stack-name serverless-api-backend-test-4 |grep FAILED
            _"EventId": "Table-CREATE_FAILED-2017-05-04T18:55:55.655Z", 
            "ResourceStatus": "CREATE_FAILED", 
            "EventId": "PutFunctionRole-CREATE_FAILED-2017-05-04T18:55:55.032Z", 
            "ResourceStatus": "CREATE_FAILED", 
            "EventId": "DeleteFunctionRole-CREATE_FAILED-2017-05-04T18:55:54.906Z", 
            "ResourceStatus": "CREATE_FAILED", 
            "EventId": "GetFunctionRole-CREATE_FAILED-2017-05-04T18:55:54.702Z", 
            "ResourceStatus": "CREATE_FAILED",_

**Comment by syang** [comment]
After grant this user, I called it serverlessAdmin, all permissions (essentially an account admin), the above errors goes away. But I don't think it's a good practice. Would be great if we can find out the minimal Policy that should be attached to this user.

**Comment by purefan** [comment]
I was facing the same problem and yeah adding permissions helped but ideally I would like to see an error message that describes to me what went wrong

**Comment by timjrobinson** [comment]
I just got this issue and it's because I was missing the `cloudformation:DescribeChangeSet` permission.

**Comment by sanathkr** [comment]
Nothing actionable here. Closing

**Comment by dannybster** [comment]
For anybody visiting here with a similar issue I had this problem but there were no Events in Cloudformation alerting me to missing permissions. I then used `aws cloudformation execute-change-set` which failed and alerted me to the missing permissions in the Cloudformation dashboard. After adding what I now believe are the minimum permissions execute-change-set works but deploy still does not.

My current understanding of minimum permissions 
```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "s3:PutObject",
                "s3:GetObject"
            ],
            "Resource": "arn:aws:s3:::your-bucket/*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "cloudformation:DescribeStacks",
                "cloudformation:ExecuteChangeSet",
                "cloudformation:GetTemplateSummary"
            ],
            "Resource": "arn:aws:cloudformation:your-region:your-account:stack/your-stack/*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "cloudformation:CreateChangeSet"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "iam:GetRole"
            ],
            "Resource": "*"
        },
        {
            "Effect": "Allow",
            "Action": [
                "lambda:UpdateFunctionCode",
                "lambda:GetFunctionConfiguration"
            ],
            "Resource": "arn:aws:lambda:your-region:your-account:function:your-function"
        }
    ]
}
```

**Comment by markdwags** [comment]
I ran into this as well, using 2.13.16.

It was failing the same way, same error of "Status". I was missing `DescribeChangeSet`

**Comment by josealdaco** [comment]
Same issue, was missing `DescribeChangeSet ` but the output gave no meaningful feedback on the source of the problem. Before I would just run the cli command a second time which would then succeed, the output should return a standard IAM role `Does not have permissions to run cloudformation:DescribeChangeSet on resource xxx`.

**Comment by tyliggity** [comment]
Guys, this was such a simple fix and yet was closed as "nothing actionable"?  The error message is `'Status'` which is NOT an error message, let alone a broken one.  Terrible that I have to go back to this 6 year old issue to find the missing permission.

**Comment by GavinZZ** [comment]
The reason we marked it as `nothing actionable` is because this is using AWS CLI, which is not in our control. We would recommend using [SAM CLI](https://docs.aws.amazon.com/serverless-application-model/latest/developerguide/serverless-getting-started-hello-world.html) to deploy templates, i.e. `sam deploy --guided`.

**Comment by tyliggity** [comment]
@GavinZZ I understand.  At the very least, this should be escalated to the AWS CLI folks then.  SAM CLI is surely not the recommended approach to deploying CloudFormation templates which have nothing to do with serverless solutions.

**Comment by GavinZZ** [comment]
@tyliggity Let me rephrase. If the template uses SAM Transform, it’s definitely recommended to use SAM CLI. However, if it’s a pure CloudFormation template, it definitely make sense to use AWS CLI. 

I’ll help cut an issue to aws-cli repository to provide a more meaningful error message.
